batcher/sweep_batcher.go

```diff
@@ -0,0 +1,591 @@
+// placing them in appropriate batches. It will spin up new batches as needed.
+type Batcher struct {
+	// batches is a map of batch IDs to the currently active batches.
+	batches map[int32]*batch
+
+	// SweepReqs is a channel where sweep requests are received.
+	SweepReqs chan SweepRequest
+
+	// errChan is a channel where errors are received.
+	errChan chan error
+
+	// completeChan is a channel where batches signal their completion by
+	// providing their ID.
+	completeChan chan int32
+
+	// wallet is the wallet kit client that is used by batches.
+	wallet lndclient.WalletKitClient
+
+	// chainNotifier is the chain notifier client that is used by batches.
+	chainNotifier lndclient.ChainNotifierClient
+
+	// signerClient is the signer client that is used by batches.
+	signerClient lndclient.SignerClient
+
+	// musig2ServerKit includes all the required functionality to collect
+	// and verify signatures by the swap server in order to cooperatively
+	// sweep funds.
+	musig2ServerSign MuSig2SignSweep
+
+	// verifySchnorrSig is a function that can be used to verify a schnorr
+	// signature.
+	VerifySchnorrSig VerifySchnorrSig
+
+	// chainParams are the chain parameters of the chain that is used by
+	// batches.
+	chainParams *chaincfg.Params
+
+	// store includes all the database interactions that are needed by the
+	// batcher and the batches.
+	store BatcherStore
+
+	// wg is a waitgroup that is used to wait for all the goroutines to
+	// exit.
+	wg sync.WaitGroup
+}
+
+// NewBatcher creates a new Batcher instance.
+func NewBatcher(wallet lndclient.WalletKitClient,
+	chainNotifier lndclient.ChainNotifierClient,
+	signerClient lndclient.SignerClient, musig2ServerSigner MuSig2SignSweep,
+	verifySchnorrSig VerifySchnorrSig, chainparams *chaincfg.Params,
+	store BatcherStore) *Batcher {
+
+	return &Batcher{
+		batches:          make(map[int32]*batch),
+		SweepReqs:        make(chan SweepRequest),
+		errChan:          make(chan error),
+		completeChan:     make(chan int32),
+		wallet:           wallet,
+		chainNotifier:    chainNotifier,
+		signerClient:     signerClient,
+		musig2ServerSign: musig2ServerSigner,
+		VerifySchnorrSig: verifySchnorrSig,
+		chainParams:      chainparams,
+		store:            store,
+	}
+}
+
+// Run starts the batcher and processes incoming sweep requests.
+func (b *Batcher) Run(ctx context.Context) error {
+	runCtx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+
+		for _, batch := range b.batches {
+			batch.Wait()
+		}
+
+		b.wg.Wait()
+	}()
+
+	// First we fetch all the batches that are not in a confirmed state from
+	// the database. We will then resume the execution of these batches.
+	batches, err := b.FetchUnconfirmedBatches(runCtx)
+	if err != nil {
+		return err
+	}
+
+	for _, batch := range batches {
+		err := b.spinUpBatchFromDB(runCtx, batch)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case sweepReq := <-b.SweepReqs:
```

Comment:
Instead of using the exposed `SweepReqs` channel, I'd recommend adding an `AddSweep` member function instead. Exposing the bare channel can cause problems if we were to ever wanted wait for the swap to exit (and the batcher already quit).